File: EstruturasDeDados/pilhaflexivel/pilhaflexivel.c
#include <stdio.h>
#include <stdlib.h>

typedef struct Celula {
  int elemento;
  struct Celula *prox;
} Celula;

typedef struct {
  Celula *topo;
} Pilha;

void inserir(Pilha *pilha, int x) {
  Celula *tmp = (Celula *)malloc(sizeof(Celula));
  if (pilha->topo == NULL) {
    pilha->topo = tmp;
    tmp->elemento = x;
    tmp->prox = NULL;
  } else {
    tmp->prox = pilha->topo;
    pilha->topo = tmp;
    tmp->elemento = x;
  }
}

int remover(Pilha *pilha) {
  if (pilha->topo == NULL) {
    printf("Vazia!");
    return -1;
  }
  int el;
  Celula *tmp = pilha->topo;
  pilha->topo = tmp->prox;
  el = tmp->elemento;
  free(tmp);
  return el;
}

void mostrar(Pilha *pilha) {
  Celula *tmp = pilha->topo;

  printf("[");
  while (tmp != NULL) {
    printf("%d ", tmp->elemento);
    tmp = tmp->prox;
  }
  printf("]");
}
int main() {
  Pilha *pilha = (Pilha *)malloc(sizeof(Pilha));
  pilha->topo = NULL;
  inserir(pilha, 1);
  inserir(pilha, 2);
  inserir(pilha, 3);
  inserir(pilha, 4);
  mostrar(pilha);
}

File: EstruturasDeDados/pilhaflexivel/PilhaFlexivel.java
class Celula {
  int elemento;
  Celula prox;
}

class Pilha {
  Celula topo;

  Pilha() {
    topo = null;
  }

  public void inserir(int x) {
    Celula tmp = new Celula();

    if (topo == null) {
      topo = tmp;
      topo.elemento = x;
    } else {
      tmp.prox = topo;
      topo = tmp;
      tmp.elemento = x;
    }
  }

  public int remover() {
    if (topo == null) {
      System.out.println("Vazia!");
      return -1;
    } else {
      Celula tmp = topo;
      int el = tmp.elemento;
      topo = topo.prox;
      tmp = tmp.prox = null;
      return el;
    }
  }

  public void mostrar() {
    if (topo == null) {
      System.out.println("Vazia!");
      return;
    }

    Celula tmp = topo;
    System.out.print("[");
    while (tmp != null) {
      System.out.print(tmp.elemento + " ");
      tmp = tmp.prox;
    }
    System.out.print("]");
  }
}

class PilhaFlexivel {
  public static void main(String[] args) {
    Pilha pilha = new Pilha();

    pilha.inserir(1);
    pilha.inserir(2);
    pilha.inserir(3);
    pilha.inserir(4);
    System.out.println(pilha.remover());
    System.out.println(pilha.remover());
    System.out.println(pilha.remover());
    System.out.println(pilha.remover());
    System.out.println(pilha.remover());
  }
}
